File: sdk/flutter/iapstack/test/openapi_contract_test.dart
import 'dart:io';

import 'package:test/test.dart';
import 'package:yaml/yaml.dart';

void main() {
  group('OpenAPI v1 compatibility', () {
    late YamlMap document;

    setUpAll(() async {
      final contract = File('../../../contracts/openapi/v1.yaml');
      expect(await contract.exists(), isTrue,
          reason: 'The canonical repository contract must be available.');
      document = loadYaml(await contract.readAsString()) as YamlMap;
    });

    test('declares every operation used by the provider-neutral client', () {
      final paths = document['paths'] as YamlMap;
      expect(
        _operationId(
          paths,
          '/v1/applications/{application_id}/purchases:verify',
          'post',
        ),
        'verifyPurchase',
      );
      expect(
        _operationId(
          paths,
          '/v1/applications/{application_id}/purchases:restore',
          'post',
        ),
        'restorePurchases',
      );
      expect(
        _operationId(
          paths,
          '/v1/applications/{application_id}/customers/{external_customer_id}/entitlements',
          'get',
        ),
        'getCustomerEntitlements',
      );
    });

    test('keeps request, response, and error fields consumed by the SDK', () {
      final schemas = (document['components'] as YamlMap)['schemas'] as YamlMap;
      _expectRequiredFields(schemas, 'PurchaseSubmission', const <String>{
        'external_customer_id',
        'claimed_products',
        'evidence',
      });
      _expectRequiredFields(schemas, 'VerificationResult', const <String>{
        'verified_at',
        'customer_id',
        'entitlements',
      });
      _expectRequiredFields(
          schemas, 'RestoreResult', const <String>{'results'});
      _expectRequiredFields(schemas, 'EntitlementSnapshot', const <String>{
        'customer_id',
        'entitlements',
      });
      _expectRequiredFields(schemas, 'ErrorEnvelope', const <String>{'error'});
      _expectRequiredFields(schemas, 'ApiError', const <String>{
        'code',
        'message',
      });
    });
  });
}

String _operationId(YamlMap paths, String path, String method) {
  final pathItem = paths[path] as YamlMap;
  final operation = pathItem[method] as YamlMap;
  return operation['operationId'] as String;
}

void _expectRequiredFields(
  YamlMap schemas,
  String schemaName,
  Set<String> expected,
) {
  final schema = schemas[schemaName] as YamlMap;
  final properties = schema['properties'] as YamlMap;
  final required = (schema['required'] as YamlList).cast<String>().toSet();
  expect(properties.keys.cast<String>(), containsAll(expected));
  expect(required, containsAll(expected));
}

File: sdk/android/core/src/test/kotlin/com/iapstack/core/OpenApiContractTest.kt
package com.iapstack.core

import org.yaml.snakeyaml.Yaml
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class OpenApiContractTest {
  private val document: Map<String, Any?> = loadContract()

  @Test
  fun declaresEveryOperationUsedByTheProviderNeutralClient() {
    val paths = mapAt(document, "paths")
    assertEquals("verifyPurchase", operationId(paths, "/v1/applications/{application_id}/purchases:verify", "post"))
    assertEquals("restorePurchases", operationId(paths, "/v1/applications/{application_id}/purchases:restore", "post"))
    assertEquals(
      "getCustomerEntitlements",
      operationId(
        paths,
        "/v1/applications/{application_id}/customers/{external_customer_id}/entitlements",
        "get",
      ),
    )
  }

  @Test
  fun keepsRequestResponseAndErrorFieldsConsumedByTheSdk() {
    val schemas = mapAt(mapAt(document, "components"), "schemas")
    expectRequiredFields(schemas, "PurchaseSubmission", setOf("external_customer_id", "claimed_products", "evidence"))
    expectRequiredFields(schemas, "VerificationResult", setOf("verified_at", "customer_id", "entitlements"))
    expectRequiredFields(schemas, "RestoreResult", setOf("results"))
    expectRequiredFields(schemas, "EntitlementSnapshot", setOf("customer_id", "entitlements"))
    expectRequiredFields(schemas, "ErrorEnvelope", setOf("error"))
    expectRequiredFields(schemas, "ApiError", setOf("code", "message"))
  }

  private fun operationId(paths: Map<String, Any?>, path: String, method: String): String {
    val pathItem = mapAt(paths, path)
    val operation = mapAt(pathItem, method)
    return operation["operationId"] as String
  }

  private fun expectRequiredFields(schemas: Map<String, Any?>, schemaName: String, expected: Set<String>) {
    val schema = mapAt(schemas, schemaName)
    val properties = mapAt(schema, "properties").keys
    val required = (schema["required"] as List<*>).map { it as String }.toSet()
    assertTrue(properties.containsAll(expected), "$schemaName properties missing $expected")
    assertTrue(required.containsAll(expected), "$schemaName required missing $expected")
  }

  private fun mapAt(source: Map<String, Any?>, key: String): Map<String, Any?> {
    @Suppress("UNCHECKED_CAST")
    return source[key] as Map<String, Any?>
  }

  private fun loadContract(): Map<String, Any?> {
    val contract = File("../../../contracts/openapi/v1.yaml")
    assertTrue(contract.isFile, "The canonical repository contract must be available.")
    @Suppress("UNCHECKED_CAST")
    return Yaml().load<Map<String, Any?>>(contract.readText())
  }
}
